src/khepri_lock.erl

```diff
@@ -0,0 +1,654 @@
+  #state{locks = Locks, monitors = Monitors} = State) ->
+    case Locks of
+        #{ResourceId := {GroupId, Acquirers}} ->
+            %% Remove only the first found acquisition by this process.
+            case lists:keytake(Pid, 1, Acquirers) of
+                {value, {Pid, MonitorRef}, Acquirers1} ->
+                    case MonitorRef of
+                        undefined ->
+                            ok;
+                        _Ref ->
+                            erlang:demonitor(MonitorRef, [flush])
+                    end,
+                    Locks1 = case Acquirers1 of
+                                 [] ->
+                                     maps:remove(ResourceId, Locks);
+                                 _Others ->
+                                     LockEntry = {GroupId, Acquirers1},
+                                     Locks#{ResourceId => LockEntry}
+                             end,
+                    Monitors1 = maps:remove(MonitorRef, Monitors),
+                    State1 = State#state{locks = Locks1, monitors = Monitors1},
+                    {reply, ok, State1};
+                false ->
+                    {reply, error, State}
+            end;
+        _ ->
+            {reply, error, State}
+    end;
+handle_call({delete, ResourceId}, _From, #state{locks = Locks} = State) ->
+    case Locks of
+        #{ResourceId := {_GroupId, Acquirers}} ->
+            Locks1 = maps:remove(ResourceId, Locks),
+            Monitors1 = lists:foldl(
+                          fun ({_Pid, undefined}, Monitors) ->
+                                  Monitors;
+                              ({_Pid, MonitorRef}, Monitors) ->
+                                  erlang:demonitor(MonitorRef, [flush]),
+                                  maps:remove(MonitorRef, Monitors)
+                          end, State#state.monitors, Acquirers),
+            State1 = State#state{locks = Locks1, monitors = Monitors1},
+            {reply, {ok, length(Acquirers)}, State1};
+        _ ->
+            {reply, {ok, 0}, State}
+    end;
+handle_call(Request, From, State) ->
+    ?LOG_WARNING(
+      "Unhandled handle_call request from ~0p: ~p",
+      [From, Request]),
+    {noreply, State}.
+
+%% @private
+handle_cast(Request, State) ->
+    ?LOG_WARNING("Unhandled handle_cast request: ~p", [Request]),
+    {noreply, State}.
+
+%% @private
+handle_info(
+  {'DOWN', MonitorRef, process, _Pid, _Reason},
+  #state{locks = Locks, monitors = Monitors} = State) ->
+    case maps:take(MonitorRef, Monitors) of
+        {ResourceId, Monitors1} ->
+            State1 = State#state{monitors = Monitors1},
+            case Locks of
+                #{ResourceId := {GroupId, Acquirers}} ->
+                    %% Remove all acquisitions by this process.
+                    Acquirers1 = lists:filter(
+                                   fun({_, Ref}) -> Ref =/= MonitorRef end,
+                                   Acquirers),
```

Comment:
Currently, the code uses several monitors for the same process, so this loop will only remove a single acquirer, not all for this process. There is no bug though because the gen_server will receive multiple `DOWN` signals.